### PR TITLE
feat: update output SAM flags and tags

### PR DESCRIPTION
**Pull request by nh13**
This improves and standardizes how to identify to which linear alignment chain the SAM alignment belongs, the order of the alignments in a chain, and which chain is the "primary" chain.

Implements #44.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -84,11 +85,59 @@ Since the alignment may jump to a previous reference position, different segment
...
+3. For "secondary chains", have one sub-alignment be representative.  Set the secondary
+   flag (since it's not in the "primary chain") but do not set the supplementary flag.
+4. For "secondary chains", have the non-representative sub-alignments have both the
+   secondary and supplementary flags set.
+
+See `--pick-primary` for how the primary alignment is chosen (and so when the secondary flag is 
+set for the other alignments in the given chain).
+
+The SAM tags are set as follows:
+- qs: the zero-based index of the first query base in the sub-alignment
```
Comment (nh13) [inline README.md:129]: is a table better than a list here?

File: fg-stitch-lib/src/align/aligners/mod.rs [hunk 2]
```diff
@@ -46,7 +46,7 @@ use noodles::{
         alignment::Record as SamRecord,
         record::{
             cigar::op::{Kind, Op},
-            data::field::tag::{ALIGNMENT_HIT_COUNT, ALIGNMENT_SCORE, HIT_INDEX, TOTAL_HIT_COUNT},
```
Comment (nh13) [inline fg-stitch-lib/src/align/aligners/mod.rs:49]: I find the NH/IH/HI tags confusing when present.  In particular that (1) the NH tag can be > than the # of alignments in the file, (2) that HI is the index of the alignment in the file (should change if we reorder the file).  So removing them.

File: fg-stitch-lib/src/align/aligners/mod.rs [hunk 3]
```diff
@@ -626,15 +626,16 @@ impl<'a, F: MatchFunc> SamRecordFormatter<'a, F> {
     pub fn format(
         &self,
         fastq: &FastqOwnedRecord,
-        alignments: &[Alignment],
-        alt_score: Option<i32>,
+        chains: &[Alignment],
+        pre_alignment_score: Option<i32>,
```
Comment (nh13) [inline fg-stitch-lib/src/align/aligners/mod.rs:630]: updated names to be closer than what the represent.  Perhaps we should consider (in another issue), changing the `Alignment` class to be `Chain` or the like?

File: fg-stitch-lib/src/align/aligners/mod.rs [hunk 4]
```diff
@@ -667,48 +668,110 @@ impl<'a, F: MatchFunc> SamRecordFormatter<'a, F> {
             return Ok([record].to_vec());
         }
 
+        // A note on how SAM flags are set
```
Comment (nh13) [inline fg-stitch-lib/src/align/aligners/mod.rs:671]: while redundant (and mostly copied verbatim in the README), it is useful to detail here again for the developer.  The only worry is maintaining the same information in two places.  would we prefer to just link to the README here instead?

File: fg-stitch-lib/src/align/aligners/mod.rs [hunk 5]
```diff
@@ -838,29 +910,33 @@ impl<'a, F: MatchFunc> SamRecordFormatter<'a, F> {
...
                 );
                 data.insert(
-                    TOTAL_HIT_COUNT,
-                    noodles::sam::record::data::field::Value::from(subs.len() as i32),
+                    "cn".parse().unwrap(),
+                    noodles::sam::record::data::field::Value::from(chains.len() as i32),
+                );
+                data.insert(
+                    ALIGNMENT_SCORE,
+                    noodles::sam::record::data::field::Value::from(sub.score),
```
Comment (nh13) [inline fg-stitch-lib/src/align/aligners/mod.rs:939]: I think we should add some unit tests for these tags.  Perhaps we spot check some alignments manually, then create an issue to add tests!

File: README.md [hunk 6]
```diff
@@ -84,11 +85,59 @@ Since the alignment may jump to a previous reference position, different segment
...
+are set as follows:
+1. One sub-alignment will (a) not have the secondary flag set, and (b) not have the
+   supplementary flag set.  This is the "primary" (representative) sub-alignment in the
+   "primary chain" (NB: not a formal term).
+2. The remaining sub-alignments in the "primary chain" have the supplementary flag set,
+   but not the secondary flag set.  They are part of the best linear alignment chain,
+   but not the representative sub-alignment.
+3. For "secondary chains", have one sub-alignment be representative.  Set the secondary
+   flag (since it's not in the "primary chain") but do not set the supplementary flag.
+4. For "secondary chains", have the non-representative sub-alignments have both the
```
Comment (jdidion) [inline README.md:122]:
```suggestion
4. For "secondary chains", the non-representative sub-alignments have both the
```

File: README.md [hunk 7]
```diff
@@ -84,11 +85,59 @@ Since the alignment may jump to a previous reference position, different segment
...
+   secondary and supplementary flags set.
+
+See `--pick-primary` for how the primary alignment is chosen (and so when the secondary flag is
```
Comment (jdidion) [inline README.md:125]: Not sure what exactly you mean in the parenthetical statement - consider revising

File: README.md [hunk 8]
```diff
@@ -84,11 +85,59 @@ Since the alignment may jump to a previous reference position, different segment
...
+- ts: the zero-based index of the first target base in the sub-alignment
+- te: the zero-based exclusive index of the last target base in the sub-alignment
+- as: the alignment score of the chain (not the sub-alignmnet, see AS for that)
+- xs: the sub-optimal alignment score, practically the maximum of any pre-alignment and
+      secondary chain.
+- si: the index of the sub-alignment in the current chain
+- cl: the number of sub-alignments in the current chain
+- ci: the index of the chain across all chains for this query
+- cn: the number of chains for this query
+- AS: the alignment score of the sub-alignment (not the chain, see as for that)
```
Comment (jdidion) [inline README.md:140]:
```suggestion
- AS: the alignment score of the sub-alignment (not the chain, see `as` for that)
```

File: README.md [hunk 9]
```diff
@@ -84,11 +85,59 @@ Since the alignment may jump to a previous reference position, different segment
...
+
+See `--pick-primary` for how the primary alignment is chosen (and so when the secondary flag is 
+set for the other alignments in the given chain).
+
+The SAM tags are set as follows:
+- qs: the zero-based index of the first query base in the sub-alignment
+- qe: the zero-based exclusive index of the last query base in the sub-alignment
+- ts: the zero-based index of the first target base in the sub-alignment
+- te: the zero-based exclusive index of the last target base in the sub-alignment
+- as: the alignment score of the chain (not the sub-alignmnet, see AS for that)
```
Comment (jdidion) [inline README.md:133]:
```suggestion
- as: the alignment score of the chain (not the sub-alignmnet, see `AS` for that)
```

File: fg-stitch-lib/src/align/aligners/mod.rs (hunk 2 above)
Comment (jdidion) [reply fg-stitch-lib/src/align/aligners/mod.rs:49]: I think `NH` is for the scenario where the aligner is configured to only emit a subset of alignments. But I agree they are not as clear/useful as the tags you've added. I would only add them if there were a downstream tool that expected them, but we can wait for someone to open an issue.

File: fg-stitch-lib/src/align/aligners/mod.rs (hunk 3 above)
Comment (jdidion) [reply fg-stitch-lib/src/align/aligners/mod.rs:630]: Yes I think `Chain` or `AlignmentChain` would be better.

File: fg-stitch-lib/src/align/aligners/mod.rs [hunk 10]
```diff
@@ -667,48 +668,110 @@ impl<'a, F: MatchFunc> SamRecordFormatter<'a, F> {
...
+            if chain_index == 0 {
+                primary_alignment_score = subs[primary_sub_idx].score;
+            }
+
             // Filter out sub-alignments that have score worse than X% of the primary
-            // if opts.  This changes the primary index!
+            // if specified.  This may change the primary index for this chain!
             if self.opts.filter_secondary {
-                let min_score = primary_score as f32 * self.opts.filter_secondary_pct / 100.0;
+                // Get the minimum alingmnet score to keep.
```
Comment (jdidion) [inline fg-stitch-lib/src/align/aligners/mod.rs:764]:
```suggestion
                // Get the minimum alignment score to keep.
```

File: fg-stitch-lib/src/align/aligners/mod.rs [hunk 11]
```diff
@@ -838,29 +910,33 @@ impl<'a, F: MatchFunc> SamRecordFormatter<'a, F> {
...
+                if let Some(score) = suboptimal_score {
                     data.insert(
                         "xs".parse().unwrap(),
                         noodles::sam::record::data::field::Value::from(score),
                     );
                 }
                 data.insert(
-                    ALIGNMENT_SCORE,
-                    noodles::sam::record::data::field::Value::from(sub.score),
+                    "si".parse().unwrap(),
```
Comment (jdidion) [inline fg-stitch-lib/src/align/aligners/mod.rs:922]: Introduce constants for tag names

File: README.md (hunk 1 above)
Comment (jdidion) [reply README.md:129]: Yes I'd go with a table

File: README.md (hunk 7 above)
Comment (nh13) [reply README.md:125]: Just removed as it is confusing, and duplicates the description above!

File: fg-stitch-lib/src/align/aligners/mod.rs [hunk 12]
```diff
@@ -667,48 +668,110 @@ impl<'a, F: MatchFunc> SamRecordFormatter<'a, F> {
...
+
             // Filter out sub-alignments that have score worse than X% of the primary
-            // if opts.  This changes the primary index!
+            // if specified.  This may change the primary index for this chain!
             if self.opts.filter_secondary {
-                let min_score = primary_score as f32 * self.opts.filter_secondary_pct / 100.0;
+                // Get the minimum alingmnet score to keep.
+                let min_score =
+                    primary_alignment_score as f32 * self.opts.filter_secondary_pct / 100.0;
                 let mut new_subs = Vec::new();
```
[the remark this answers is not in the document]
Comment (nh13) [reply fg-stitch-lib/src/align/aligners/mod.rs:762]: Tried to get this done with:

```rust
        let (new_subs, _) = subs.iter().fold(
                    (Vec::with_capacity(subs.len()), 0),
                    |(mut new_subs, old_idx), sub| {
                        if old_idx == primary_sub_idx {
                            primary_sub_idx = new_subs.len();
                        }
                        if sub.score as f32 >= min_score {
                            new_subs.push(sub);
                        }
                        (new_subs, old_idx + 1)
                    },
                );
                subs = new_subs;
``` 
but `new_subs` has type `Vec<&SubAlignment>`, but we need `Vec<SubAlignment>`.  Will defer.

File: fg-stitch-lib/src/align/aligners/mod.rs [hunk 13]
```diff
@@ -30,6 +30,11 @@ use crate::{
     util::{
         dna::reverse_complement,
         index_map::IndexMap,
+        tag::{
```
Comment (jdidion) [inline fg-stitch-lib/src/align/aligners/mod.rs:33]: Fine for now, but looking at it again, this is probably a place where you want to introduce eg `enum CustomTags` and implement `From<CustomTag> for Tag`.

**Review by jdidion — APPROVED** [review]

**Pull request merged**